### PR TITLE
Feat/hyperparameter tuning

### PR DESCRIPTION
**Pull request by linus-mt**

### CODE REVIEW

File: sotai/api.py [hunk 1]
```diff
@@ -351,12 +358,102 @@ def get_inference_results(inference_uuid: str, download_folder: str) -> APIStatu
...
     return APIStatus.SUCCESS
+
+
+def post_dataset(
+    data_filepath: str,
+    columns: List[str],
+    categorical_columns: List[str],
+    pipeline_uuid: str,
+) -> Tuple[APIStatus, Optional[str]]:
+    """Upload a dataset to th the SOTAI API .
```
Comment (willbakst) [inline sotai/api.py:378]: nit: remove extra space before period.

File: sotai/api.py [hunk 2]
```diff
@@ -351,12 +358,102 @@ def get_inference_results(inference_uuid: str, download_folder: str) -> APIStatu
...
+
+    Args:
+        data_filepath: The path to the data file to create the inference for.
+        pipeline_uuid: The pipeline uuid to upload the dataset for.
+
+    Returns:
+        A tuple containing the status of the API call and the UUID of the created
+        inference job. If unsuccessful, the UUID will be None.
```
Comment (willbakst) [inline sotai/api.py:386]: nit: update None -> `None`

File: sotai/api.py [hunk 3]
```diff
@@ -351,12 +358,102 @@ def get_inference_results(inference_uuid: str, download_folder: str) -> APIStatu
...
+def post_dataset(
+    data_filepath: str,
+    columns: List[str],
+    categorical_columns: List[str],
+    pipeline_uuid: str,
+) -> Tuple[APIStatus, Optional[str]]:
+    """Upload a dataset to th the SOTAI API .
+
+    Args:
+        data_filepath: The path to the data file to create the inference for.
```
Comment (willbakst) [inline sotai/api.py:381]: is this supposed to say inference?

File: sotai/api.py [hunk 4]
```diff
@@ -351,12 +358,102 @@ def get_inference_results(inference_uuid: str, download_folder: str) -> APIStatu
...
+        pipeline_uuid: The pipeline uuid to upload the dataset for.
```
Comment (willbakst) [inline sotai/api.py:382]: nit: update to "The pipeline UUID for which to upload the dataset.

File: sotai/api.py [hunk 5]
```diff
@@ -351,12 +358,102 @@ def get_inference_results(inference_uuid: str, download_folder: str) -> APIStatu
...
+    pipeline_config: PipelineConfig,
+    dataset_uuid: str,
+):
+    """Upload a dataset to th the SOTAI API .
+
+    Args:
+        hypertune_config: The hypertune config to create the hypertune job for.
+        pipeline_config: The pipeline config to create the hypertune job for.
+        dataset_uuid: The dataset uuid to create the hypertune job for.
+
```
Comment (willbakst) [inline sotai/api.py:419]: remove extra newline here

File: sotai/api.py [hunk 6]
```diff
@@ -351,12 +358,102 @@ def get_inference_results(inference_uuid: str, download_folder: str) -> APIStatu
...
+
+    Returns:
+        A tuple containing the status of the API call and the UUID of the created
+        inference job. If unsuccessful, the UUID will be None.
+
```
Comment (willbakst) [inline sotai/api.py:424]: remove extra newline here

File: sotai/api.py [hunk 7]
```diff
@@ -351,12 +358,102 @@ def get_inference_results(inference_uuid: str, download_folder: str) -> APIStatu
...
+    return APIStatus.SUCCESS, response.json()["uuid"]
+
+
+def post_hypertune_job(
+    hypertune_config: HypertuneConfig,
+    pipeline_config: PipelineConfig,
+    dataset_uuid: str,
+):
+    """Upload a dataset to th the SOTAI API .
```
Comment (willbakst) [inline sotai/api.py:413]: nit: remove extra space before period.

File: sotai/api.py [hunk 8]
```diff
@@ -351,12 +358,102 @@ def get_inference_results(inference_uuid: str, download_folder: str) -> APIStatu
...
+
+    Args:
+        hypertune_config: The hypertune config to create the hypertune job for.
+        pipeline_config: The pipeline config to create the hypertune job for.
+        dataset_uuid: The dataset uuid to create the hypertune job for.
+
+
+    Returns:
+        A tuple containing the status of the API call and the UUID of the created
+        inference job. If unsuccessful, the UUID will be None.
```
Comment (willbakst) [inline sotai/api.py:423]: nit: update None -> `None`

File: sotai/api.py [hunk 9]
```diff
@@ -351,12 +358,102 @@ def get_inference_results(inference_uuid: str, download_folder: str) -> APIStatu
...
+            },
+            "model_config": {
+                "model_framework": "pytorch",
+                "model_config_name": "Model 1",
+                "model_type": "linear",
+                "target_column_type": pipeline_config.target_type.value,
+                "target_column": pipeline_config.target,
+                "primary_metric": pipeline_config.primary_metric.value,
+                "selected_features": [
+                    key for key in pipeline_config.feature_configs.keys()
```
Comment (willbakst) [inline sotai/api.py:445]: unnecessary list comprehension. Instead just use `list(pipeline_config.feature_configs.keys())`

File: sotai/enums.py [hunk 10]
```diff
@@ -159,3 +159,9 @@ class InferenceConfigStatus(_Enum):
     PREPARING = "preparing"
     RUNNING = "running"
     SUCCESS = "success"
+
+
+class HypertuneMethod(_Enum):
+    """Enum for HypertuneMethod"""
```
Comment (willbakst) [inline sotai/enums.py:165]: update as follows:

```
class HypertuneMethod(_Enum):
    """Type of hyperparameter optimzation method.

    - GRID: searches the grid generated by taking the cross product of all hyperparameter options.
    """

    GRID = "grid"
```

File: sotai/pipeline.py [hunk 11]
```diff
@@ -249,6 +252,90 @@ def train(
...
+                hypertune job will be run locally.
+
+
+        Returns:
+            A list of `TrainedModel` instances if run locally, or a list of trained
+            model uuids if run in the SOTAI cloud.
+        """
+
+        if not pipeline_config_id:
+            pipeline_config = self._version_pipeline_config(data, pipeline_config_id)
```
Comment (willbakst) [inline sotai/pipeline.py:284]: qq: this is different from train since we're not preparing the dataset. is this because we're preparing it in the cloud? I'm condering if we should prepare it just like we would for train locally but then upload the prepared dataset so that it's properly managed here like the config is when we version it?

File: sotai/api.py [hunk 12]
```diff
@@ -351,12 +358,102 @@ def get_inference_results(inference_uuid: str, download_folder: str) -> APIStatu
...
+        )
+    if response.status_code != 200:
+        logging.error("Failed to upload dataset")
+        logging.error(response.json())
+        return APIStatus.ERROR, None
+
+    return APIStatus.SUCCESS, response.json()["uuid"]
+
+
+def post_hypertune_job(
```
Comment (willbakst) [inline sotai/api.py:408]: it doesn't seem like this post request knows what model config to use?

File: sotai/pipeline.py [hunk 13]
```diff
@@ -249,6 +252,90 @@ def train(
...
+        hyperparameter_permutations = list(
+            itertools.product(
+                hypertune_config.batch_sizes,
+                hypertune_config.epochs,
+                hypertune_config.learning_rates,
+            )
+        )
+        trained_models = []
+
+        if model_config is None:
```
Comment (willbakst) [inline sotai/pipeline.py:314]: this should happen earlier even before the hosted call and should be passed to the hypertune job

File: sotai/pipeline.py [hunk 14]
```diff
@@ -249,6 +252,90 @@ def train(
...
+
+            _, dataset_uuid = self._upload_dataset(self.uuid, data)
+            hypertune_response, trained_model_uuids = post_hypertune_job(
+                hypertune_config, pipeline_config, dataset_uuid
+            )
+            if hypertune_response == APIStatus.ERROR:
+                return []
+            return trained_model_uuids
+
+        hyperparameter_permutations = list(
```
Comment (willbakst) [inline sotai/pipeline.py:305]: there's no need to make this a list since itertools.product returns an iterator. You can just iterate through it as follows:
```
hyperparameter_grid = itertools.product(...)
for i, permutation in enumerate(hyperparameter_grid):
    ...
```

File: sotai/pipeline.py [hunk 15]
```diff
@@ -249,6 +252,90 @@ def train(
...
+                hypertune_config.epochs,
+                hypertune_config.learning_rates,
+            )
+        )
+        trained_models = []
+
+        if model_config is None:
+            model_config = LinearConfig()
+
+        for i, permutation in enumerate(hyperparameter_permutations):
```
Comment (willbakst) [inline sotai/pipeline.py:317]: nit: update permutation -> hyperparameters

File: sotai/pipeline.py [hunk 16]
```diff
@@ -527,3 +594,61 @@ def _upload_model(
...
+    ):
+        filepath = "/tmp/sotai"
+        if not os.path.exists(filepath):
+            os.makedirs("/tmp/sotai/")
+        dataset.to_csv("/tmp/sotai/dataset.csv")
+        columns = dataset.columns.tolist()
+        dataset_response, dataset_uuid = post_dataset(
+            "/tmp/sotai/dataset.csv",
+            columns=columns,
+            categorical_columns=["thal"],
```
Comment (willbakst) [inline sotai/pipeline.py:611]: this looks wrong?

File: sotai/pipeline.py [hunk 17]
```diff
@@ -527,3 +594,61 @@ def _upload_model(
...
+    def _post_pipeline_config(self, pipeline_config: PipelineConfig):
+        """Posts a pipeline config to the SOTAI web client. If a pipeline config has already
+        been posted, this function will return without posting the config again.
+
+        Args:
+            pipeline_config: The pipeline config to post.
+
+        Returns:
+            If the pipeline config was successfully posted, the pipeline config UUID.
+            Otherwise, None.
```
Comment (willbakst) [inline sotai/pipeline.py:626]: nit: update None -> `None`

File: sotai/pipeline.py [hunk 18]
```diff
@@ -527,3 +594,61 @@ def _upload_model(
...
+            "/tmp/sotai/dataset.csv",
+            columns=columns,
+            categorical_columns=["thal"],
+            pipeline_uuid=pipeline_uuid,
+        )
+
+        return dataset_response, dataset_uuid
+
+    def _post_pipeline_config(self, pipeline_config: PipelineConfig):
+        """Posts a pipeline config to the SOTAI web client. If a pipeline config has already
```
Comment (willbakst) [inline sotai/pipeline.py:618]: Description here can only be a single line. Reduce it to one line and push any additional information onto the next line with a newline in between i.e.

```
"""Posts a pipeline config to the SOTAI web client.

If a pipeline config has already...
"""
```

File: sotai/types.py [hunk 19]
```diff
@@ -264,3 +265,19 @@ class PipelineConfig(BaseModel):
     shuffle_data: bool = Field(...)
     drop_empty_percentage: int = Field(...)
     dataset_split: DatasetSplit = Field(...)
+
+
+class HypertuneConfig(BaseModel):
+    """A configuration object for a HypertuneConfig.
```
Comment (willbakst) [inline sotai/types.py:271]: nit: update to "A configuration object for hypertuning."

File: sotai/types.py [hunk 20]
```diff
@@ -264,3 +265,19 @@ class PipelineConfig(BaseModel):
...
+        epochs: A list of epochs to try.
+        batch_sizes: A list of batch sizes to try.
+        learning_rates: A list of learning rates to try.
+    """
+
+    loss_type: LossType
+    epochs: List[int]
+    batch_sizes: List[int]
+    learning_rates: List[float]
+    hypertune_method: HypertuneMethod = Field(default=HypertuneMethod.GRID)
```
Comment (willbakst) [inline sotai/types.py:283]: do we need this since we're only doing GRID right now?

File: sotai/types.py [hunk 21]
```diff
@@ -264,3 +265,19 @@ class PipelineConfig(BaseModel):
     shuffle_data: bool = Field(...)
     drop_empty_percentage: int = Field(...)
     dataset_split: DatasetSplit = Field(...)
+
+
+class HypertuneConfig(BaseModel):
+    """A configuration object for a HypertuneConfig.
+
+    Attributes:
+        epochs: A list of epochs to try.
```
Comment (willbakst) [inline sotai/types.py:274]: missing two attributes here. Please include them.

**Review by willbakst — COMMENTED** [review]
Majority nits, a few meatier comments

File: sotai/api.py (hunk 2 above)
Comment (linus-mt) [reply sotai/api.py:391]: What do you mean by this?

File: sotai/api.py (hunk 12 above)
Comment (linus-mt) [reply sotai/api.py:413]: I've added it.

File: sotai/pipeline.py (hunk 14 above)
Comment (linus-mt) [reply sotai/pipeline.py:308]: Yeah so we need to convert it to a list to get the length of it for logging later on, which does not work with iterators.

File: sotai/types.py (hunk 20 above)
Comment (linus-mt) [reply sotai/types.py:285]: We might as well have it defaulted for now, and I think this will tend to be the standard in the future for speed.

File: sotai/pipeline.py (hunk 16 above)
Comment (linus-mt) [reply sotai/pipeline.py]: We actually don't need to set the categorical columns since we don't do cleaning anymore, so I've set this to default to an empty array for now.

File: sotai/pipeline.py (hunk 11 above)
Comment (linus-mt) [reply sotai/pipeline.py:284]: Yeah so we handle everything in the cloud. We could have it done locally but I don't really see the benefit, since if a user is hypertuning in the hosted setup they probably aren't looking to manage it locally. We could change it to do that but it would take a fairly large number of changes.

File: sotai/api.py (hunk 2 above)
Comment (willbakst) [reply sotai/api.py:391]: oh the formatting of markdown makes this confusing. You should use backtics (`) one both sides of the None so when it gets autogenerated into markdown for the docs it looks like my previous comment.

File: sotai/pipeline.py (hunk 11 above)
Comment (willbakst) [reply sotai/pipeline.py:284]: The primary issue I see is that if shuffle_data is True, since we pass a data frame to the train method, we will prepare the data separately for each train call and thereby train the models on different datasets, which means we can't properly compare them.

At the very least we should call prepare once before iterating through the hyperparameters for local hypertuning and then pass the id to the train function (and then for hosted it's fine if we don't prepare for now, we can always update that later).

Comment (willbakst) [reply sotai/pipeline.py:284]: Unless of course the shuffle data issue is also true for hosted, in which case we should fix it there as well.

File: sotai/pipeline.py (hunk 14 above)
Comment (willbakst) [reply sotai/pipeline.py]: Ah ok this is fine then.

File: sotai/pipeline.py (hunk 16 above)
Comment (willbakst) [reply sotai/pipeline.py]: Ok sounds good.

File: sotai/types.py (hunk 20 above)
Comment (willbakst) [reply sotai/types.py:285]: sounds good

File: sotai/api.py (hunk 3 above)
Comment (willbakst) [reply sotai/api.py:384]: shouldn't this just be wrt. upload not inference?

File: sotai/pipeline.py (hunk 11 above)
Comment (linus-mt) [reply sotai/pipeline.py:284]: Ok I've set it to generate the prepared dataset first and the pass the id into train. I think this is definitely something we can improve going forward but data handling probably needs to be another EDD since there is a lot to do on that side.

File: sotai/api.py [hunk 22]
```diff
@@ -351,12 +361,115 @@ def get_inference_results(inference_uuid: str, download_folder: str) -> APIStatu
...
+
+    Args:
+        data_filepath: The path to the data file to push to the API.
+        columns: The columns of the dataset.
+        categorical_columns: The categorical columns of the dataset.
+        pipeline_uuid: The pipeline uuid for which to upload the dataset.
+
+    Returns:
+        A tuple containing the status of the API call and the UUID of the created
+        inference job. If unsuccessful, the UUID will be `None`.
```
Comment (willbakst) [inline sotai/api.py:391]: nit: comment here references inference job

File: sotai/pipeline.py [hunk 23]
```diff
@@ -249,6 +252,66 @@ def train(
...
+            hypertune_config: The config to be used for hypertuning the model.
+            pipeline_config_id: The id of the pipeline config to be used for training.
+                If not provided, the current pipeline config will be versioned and used.
+                If data is an int, this argument is ignored and the pipeline config used
+                to prepare the data with the given id will be used.
+            model_config: The config to be used for training the model. If not provided,
+                a default config will be used.
+            hosted: Whether to run the hypertune job on the SOTAI cloud. If False, the
+                hypertune job will be run locally.
+
```
Comment (willbakst) [inline sotai/pipeline.py:276]: nit: remove extra newline here

File: sotai/pipeline.py [hunk 24]
```diff
@@ -527,3 +570,114 @@ def _upload_model(
...
+        dataset_response, dataset_uuid = post_dataset(
+            "/tmp/sotai/dataset.csv",
+            columns=columns,
+            categorical_columns=[],
+            pipeline_uuid=pipeline_uuid,
+        )
+
+        return dataset_response, dataset_uuid
+
+    def _post_pipeline_config(self, pipeline_config: PipelineConfig):
```
Comment (willbakst) [inline sotai/pipeline.py:593]: add return type here `-> str`

File: sotai/pipeline.py [hunk 25]
```diff
@@ -527,3 +570,114 @@ def _upload_model(
...
+
+        return pipeline_config_uuid
+
+    def _local_hypertuning(
+        self,
+        hypertune_config: HypertuneConfig,
+        data: pd.DataFrame,
+        pipeline_config_id: Optional[int] = None,
+        model_config: Optional[LinearConfig] = None,
+    ):
```
Comment (willbakst) [inline sotai/pipeline.py:640]: add return type here `-> List[TrainedModel]`

**Review by willbakst — APPROVED** [review]
some minor nits remaining, otherwise LGTM
